storage/cache/tree_layout.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"fmt"
+
+	"github.com/google/trillian/storage"
+)
+
+const (
+	// depthQuantum defines the smallest supported subtree height, which all
+	// subtree heights must also be a multiple of.
+	//
+	// WARNING: The code below breaks if this value is not a multiple of 8,
+	// because it uses storage.NodeID byte representation directly.
+	depthQuantum = 8
+)
+
+// TreeLayout defines the mapping between tree node IDs and subtree IDs.
+type TreeLayout struct {
+	// sIndex contains stratum info for each multiple-of-depthQuantum node depth.
+	// Note that if a stratum spans multiple depthQuantum heights then it will be
+	// present in this slice the corresponding number of times.
+	// This index is used for fast mapping from node IDs to strata IDs.
+	sIndex []stratumInfo
+	// height is the height of the tree. It defines the maximal bit-length of a
+	// node ID that the tree can contain.
+	height int
+}
+
+// NewTreeLayout creates a tree layout based on the passed-in strata heights.
+func NewTreeLayout(heights []int) *TreeLayout {
+	// Compute the total tree height.
+	height := 0
+	for _, h := range heights {
+		height += h
+	}
```

Comment:
Should maybe verify the total height is sensible. If it ends up negative for example it will be used to create a slice before the > 0 checks below are applied.

Reply:
I moved the checks up a bit to address this issue.